README.md
Random Bamboo
=============


#Introduction

`Random Bamboo` (`RB`) is an ultra-fast C++ implement of the [`Random Forest`](https://www.stat.berkeley.edu/~breiman/RandomForests/cc_home.htm) algorithm. The program is designed for analyzing the large-scale [genome-wide association study](http://en.wikipedia.org/wiki/Genome-wide_association_study) (GWAS) and building predictive model on hundreds of thousands of SNPs for personalized medicine. `RB` adopts novel [data structure](http://bioinformatics.oxfordjournals.org/content/30/15/2171) and [Boolean operation method](http://bioinformatics.ust.hk/BOOST.html) to speedup the algorithm dramatically. The single-thread version is 40 times faster than the `R` package [`randomForest`](http://cran.r-project.org/web/packages/randomForest/index.html). [`OpenMP`](http://openmp.org/wp/) is used to enable parallelization feature. `RB` is thus appliable to large study with thousands or tens of thousands of individuals. I name this fast program as "Bamboo" because bamboos are some of the [fastest-growing plants](http://en.wikipedia.org/wiki/Bamboo) in the world.

#Usage

##Genotype Inputs

`RB` requires input data with format defined by [`PLINK`](http://pngu.mgh.harvard.edu/~purcell/plink/). To accelerate loading the genotype data, binary files should be generated via `PLINK` before applying `RB`. Suppose we have large `train.ped` file containing genotypes information and a mapping file `train.map`, both are described [here](http://pngu.mgh.harvard.edu/~purcell/plink/data.shtml). The following command generates three *genotype* files `train.bed`, `train.bim` and `train.fam`
```
plink --file train --out train --make-bed --noweb
```
Similarly we generate the testing data `test.bed`, `test.bim` and `test.fam` from `test.ped` and `test.map`
```
plink --file test --out test --make-bed --noweb
```

**NOTE** The three genotype files MUST share the same name.

**NOTE** The sixth column in `fam` file is the phenotype, with 1 as control and 2 as case.

**NOTE** Although six columns are in `fam` file, `RB` only uses the second (INDIVIDUAL ID) and sixth (PHENOTYPE) columns. The SEX in the fifth column is ignored. If SEX is required to be used in training, place it in covariate file `cat` (character) or covariate file `con` (dummy variable).

**NOTE** `RB` parses the input files according to their extensions. Only the file name without extension is needed when invoking `RB`.

##Basic Usage

Run the following command to train a model with SNPs only
```
bamboo --file train --out rb --mtry 5000 --ntree 10000
```
Four files `rb.cof`, `rb.err`, `rb.imp` and `rb.bam` will be saved to local, which contain the confusion matrix, oob error, Gini importance and trained model, respectively. The `bam` file is generally large, which can be muted by switching on the option `--nobam` if the trained model will not be used in future prediction.

##Covariates

If continuous and categorical covariates are available in *covariate* files `train.con` and `train.cat`, run the following command to train a model with covariates
```
bamboo --file train --cont train --cate train --out rb [...]
```
`[...]` can be any other available options to customize the training stage. The first line of the covariate files is the header and the first column is the individual IDs used for aligning the genotypes information given in genotype files. Below is an example of covariate files

```{r}
> fam <- read.table("train.fam", header = FALSE, as.is = TRUE)
> cat <- read.table("train.cat", header = TRUE, as.is = TRUE)
> con <- read.table("train.con", header = TRUE, as.is = TRUE)
> head(fam, 3)
            V1           V2 V3 V4 V5 V6
1 CG-L03-81789 CG-L03-82789  0  0  2  2
2 CG-L03-24222 CG-L03-22222  0  0  2  2
3 CG-L03-89134 CG-L03-89134  0  0  2  2
> head(cat, 3)
       GWAS_ID AGE_CAT GENDER STUDY CIGDAY_CAT CIG_CAT          QUIT_CAT
1 CG-L02-83435  51to55   MALE  PLCO     21to30 CURRENT CURRENT_NEVER_lt1
2 CG-L02-90230  61to65   MALE  ATBC     31to40 CURRENT CURRENT_NEVER_lt1
3 CG-L02-67314  66to70   MALE  PLCO     11to20 CURRENT CURRENT_NEVER_lt1
> head(con, 3)
       GWAS_ID EAGLE_EV2     PLCO_EV4     PLCO_EV5 ATBC_EV2
1 CG-L02-83435         0 -0.025857644  0.010665938        0
2 CG-L02-90230         0 -0.017074167 -0.066994786        0
3 CG-L02-67314         0  0.003454853 -0.150291391        0
```

**NOTE** We can create dummy variables for categorical covariates and treat them as ordinal covariates. In that case, only `con` file is needed. However, it is computationally more efficient if you simply place categorical covariates in `cat` file, since we have special data structure to accelerate the process of discrete data.  

**NOTE** If the individual contained in the three files are not the same, `RB` will find the intersection and train the the model on it.

##Prediction

We can predict testing data after training a model
```
bamboo --file train --cont train --cate train --out rb --pred test [...]
```
We can also use single bam file (`rb.bam`) saved in local to predict
```
bamboo --pred test --bam rb --out rb
```
If multiple `bam` files are save in a directory `./path`, we can use all of them to make a prediction
```
bamboo --pred test --bam ./path --out rb
```
`RB` will parse the meaning of `--bam` automatically. The feature is useful as we can run `RB` on multiple unparallelizable nodes, each with option `--nthread` enabled and different random seed set by `--seed`, and then all outputed `bam` files are used together in prediction.

**NOTE** The prediction results are saved in `prd` file.

**NOTE** Genotype files are always needed in prediction. If the model is trained with covariates, then `con` and `cat` files are needed as well.

**NOTE** The names of covariate files can be different from each other in training stage. They can even be different from the names of genotype files. For example, the following command works fine
```
bamboo --file geno_train --cont cont_covar --cate cate_covar --out rb [...]
```
However, when we are using the outputed `rb.bam` to predict new data, the covariate files and genotype files MUST share the same name. In the example below, all the data files share the name `test` as only the option `--pred` can be used to specify the data in predicting stage
```
bamboo --bam rb --pred test --out rb
```

##Using Part of Samples

Sometimes we prefer to use a subset of data in training or predicting. For example, all available genotype data are imputed and saved in the same genotype files; or all covariates are put in the same covariate files. We can use option `--trainid` to specify which individuals are going to be used in training, and use option `--testid` to specify those used in predicting. For example, we have data files `all_data.fam`, `all_data.bim`, `all_data.bed`, `all_data.con`, `all_data.cat` and two files `train_id.iid` and `test_id.iid`. Both of the `iid` files contain one column of individual IDs and no header is needed. They look like
```
CG-L02-61545
CG-L02-03491
CG-L02-67300
... ...
```
The following command builds model with individuals listed in `train_id.iid` only
```
bamboo --file all_data --cont all_data --cate all_data --trainid train_id --out rb [...]
```
With the model (`rb.bam`) fitted by the command above, we can predict the data specified in `test_id.iid`
```
bamboo --bam rb --pred all_data --testid test_id
```
Or we can do the above in one line
```
bamboo --file all_data --cont all_data --cate all_data --trainid train_id --pred all_data --testid test_id --out rb [...]
```

**NOTE** The file specified by the options `--trainid` and `--testid` must have extention `iid`.

##Weighted Sampling

The `Random Forest` algorithm creates a bootstrap data when building each tree in the forest. By default, all the training samples are equally weighted. This is generally not the case in genetics studies. For example, the training dataset can consist of samples from multiple studies, each with different sampling designs (e.g., [prospective cohort study](http://en.wikipedia.org/wiki/Prospective_cohort_study) and [retrospective case-control study](http://en.wikipedia.org/wiki/Case-control_study)). To build an unbiased predicitve model, one needs to find out the sampling weight of each individual to "recover" the real cohort. A sample weight file, which has extension `swt`, can be provided by the users to `RB`. The bootstrap data will be generated by sampling from the weights with replacement. 
```
CG-L02-61545            1
CG-L02-03491            1.2
CG-L02-67300            4.7
...
```
The individuals with positive ***relative*** weights in, e.g., `train.swt`, and those included in file specified by `--trainid` (if any) are involved in training. 
```
bamboo --file all_data --trainid train_id --swt train [...]
```
If `--swt` is not invoked, then all individuals or those specified by `--trainid` (if any) are used with equal weights.

##Using a Subset of SNPs

As suggested by Leo Breiman in [here](https://www.stat.berkeley.edu/~breiman/RandomForests/cc_home.htm#varimp)

> ... If the number of variables is very large, forests can be run once with all the variables, then run again using only the most important variables from the first run ... 

We can train a forest using specified markers only by using the option `--snpid`. A file with extension `sid` is needed, in which all SNPs to be used in training a model are listed in one column.
```
bamboo --file train --snpid snp [...]
```
The `snp.sid` doesn't have a header and looks like
```
rs20112
rs320078
rs213980
...
```

##Model Information

To print summary information of given bam file(s)
```
bamboo --bam rb
bamboo --bam ./path
```


#Options


* `-f`, `--file`. File names of three PLINK-formated genotype files, with extensions `bed`, `bim` and `fam`. Character. No default.
* `-c`, `--cont`. File name of continuous or ordinal covariates, with extension `con`. The first row is header. The first column contains individual IDs. Character. No default.
* `-a`, `--cate`. File name of categorical covariates, with extension `cat`. The first row is header. The first column contains individual IDs. Character. No default.
* `-o`, `--out`. File name of all output files. Character. If unspecified, it is set by `--file` or `--pred`.
* `-p`, `--pred`. File names of datasets used in prediction, including `bed`, `bim`, `fam`, `con` and `cat`. The last two are optional and is activated if covariates are used in training model. Character. No default.
* `-b`, `--bam`. Used to specify model(s) used in prediction. It can be the name of a single `bam` file or a directory contains multiple `bam` files. Character. No default.
* `-y`, `--trainid`. File name of individual IDs used in training model. The file, with extension `iid`, contains one column and no header. Character. If unspecified, all individuals in the datasets are used.
* `-z`, `--testid`. File name of individual IDs used in prediction. If `--file` is specified, the individuals are predicted with model trained from the specified datasets. If `--bam` is specified, the individuals are predicted with specified model(s). The file, with extension `iid`, contains one column and no header. Character. If unspecified, all individuals in the datasets are used.
* `-W`, `--swt`. File name of sampling weights. The file, with extension `swt`, contains two columns and no header. The first column contains individual IDs, and the second column contains relative sample weights. Only individuals with positive weights are involved in training. Will be used combined with `--trainid`. If unspecified, all individuals are used with equal weights. Character. No default.
* `-S`, `--snpid`. File name of SNPs' names used in training model. The file, with extension `sid`, contains one column and no header. Character. If unspecified, all SNPs in the training dataset are used.
* `-j`, `--varid`. File name of variables' names used as candidates of all splits. Generally these variables are known to have strong effects on the outcome. The file, with extension `vid`, contains one column and no header. Character. If unspecified, all variables are used as candidates in split by chance.
* `-t`, `--ntree`. Number of trees in the forest. Positive integer. Default: 1.
* `-m`, `--mtry`. Number of candidate variables in determining best split in each node. Positive integer. Default: sqrt of total number of variables.
* `-s`, `--seed`. Random seed. Positive integer. Default: 1.
* `-l`, `--maxnleaf`. Maximum leaves in a single tree. Positive integer. Default: 1000000.
* `-e`, `--minleafsize`. Minimum sample size in a single tree. Positive integer. Default: 1.
* `-i`, `--imp`. Types of variable importance. 1 - Gini; 2 - Permuted (Breiman & Cutler's fortran verion); 3 - Permuted (Liaw & Wiener's `R` package `randomForest`); 4 - Permuted (Raw, unrescaled); 5 - Permuted (Meng's rescaled method); 6 - print Gini and all permuted importance. Default: 1.
* `-d`, `--nthread`. Number of threads in parallelization. Positive integer. Default: maximum available number of CPUs allowed by the hardware.
* `-w`, `--classwt`. Class weight assigned to the case group. Positive double. Default: 1.0.
* `-u`, `--cutoff`. Unimplemented.
* `-g`, `--flip`. Flip the genotypes of SNPs with MAF > 0.5. Default: switched off. Switching on this option is tricky and thus should be used with caution. 
* `-x`, `--prox`. Save the proximity matrix to local file. Default: switched off. Computing proximity matrix is memory-consuming. 
* `-n`, `--noimp`. Don't save variable importances to local file. Default: switched off.
* `-B`, `--balance`. Balance the ratio of cases and controls in training data to 1. Default: switched off.
* `-r`, `--trace`. Print debug information. Valid in single-thread mode. Default: switched off.
* `-N`, `--nobam`. Don't save trained model to local file. Default: switched off.
* `-R`, `--selcovar`. Covariates are randomly selected as candidate splits, just like SNPs (not recommended). Default: switched off.
* `-T`, `--tidy`. Mute output except `bam` and `prd` file. Default: switched off.
* `-h`, `--help`. Print help information. Unimplemented. 
* `-v`, `--version`. Print version information. 


#Todo

- [ ] allow stratified sampling (e.g., STUDY)
- [x] allow specifying must-included variables (e.g., STUDY, AGE, etc.)
- [ ] optimize memory and speed further
- [x] mute output except `bam` file
- [ ] print data information (e.g., MAF, etc.)
- [ ] filter genotypes by MAF
- [ ] test AUC function with tie
- [ ] allow specifying cutoff in prediction
- [x] allow individual weights in sampling
- [ ] print help information




NEWS.md
# CHANGES IN RANDOM BAMBOO VERSION 0.5.x

## MAJOR CHANGES

- before 0.5.3, a covariate is randomly selected as a candidate split, which is not a good practice in GWAS since covariates are supposed to have significant effects on the risk of disease. In 0.5.3, all covariates become the must-included candidates by default. A new option `--selcovar` is added if we want to switch back to the strategy before 0.5.3, although it is not recommended. 
- an option, `--swt` is added to allow the users to specify sample weights for each sample. This is particular useful when the training dataset consists of samples from multiple studies with different sampling design. Only the individuals specified in both `--swt` (if any) and `--trainid` (if any) are involved in training. The individuals with zero weights are discarded, too
- two experimental options, `--neighbor` and `--searchback` introduced in v0.4.x are removed, since they have no significant effect on improving power due to the multiple-comparison issue as expected; all other major/minor changes and bug fixes in 0.4.x are inherited

## BUG FIXES

- fix a minor bug in function `PredictTestingSampleFromMultipleForest()`. It could lead to incorrect ntree

# CHANGES IN Random Bamboo VERSION 0.4.x

## MAJOR CHANGES

- discarded option `--noflip` by introducing a new one `--flip` instead; in the previous version, markers with MAF > 0.5 are flipped by default; this flag is saved in the local model and prompts `Random Bamboo` to flip markers with MAF > 0.5 in testing data. This is dangerous as the MAFs are calculated from the testing data and it can be misleading if sample size is small and MAF is close to 0.5
- two experimental options, `--neighbor` and `--searchback`, are added to explored LDs between markers

## BUG FIXES

- fixed a bug that the individuals may not be aligned correctly when loading testing data; all previous versions should not be used in predicting new dataset
- `RB` allows users to specify `--ntree` in prediction step; only specified number of bamboos are used in prediction; there is a minor bug when a single bam file is used as the trained model.
- if all markers specified by `--snpid` are not included in the data, the program may crash; it is fixed by printing a message before quiting if this happens 


src/example.R


# This script will create several files required by the random bamboo:
# (1) *.ped, *.map: These two files will be used with the program PLINK to generate files *.fam, *.bim and *.bed. The latter three files are used as inputs of random bamboo
# (2) *.con, *.cat: These two files will be used as (optional) inputs of random bamboo. They contain continuouse and categorical covariates
# (3) The files train.* will be used in training, while test.* will be used in testing

set.seed(1)

n<-2000 # total sample size
p<-15000 # number of SNPs in testing dataset
po<-10000 # number of SNPs in training dataset

# You can use smaller values for the three parameters above for convenience
# You don't need to modify others below

#sel<-sort(sample(1:p, po, FALSE))
sel<-1:po
sel.col<-sort(c(2*sel-1, 2*sel))

maf<-runif(p,.4,.4)
pr<-rep(rep(maf,each=2), times=n)

hap<-matrix(rbinom(2*n*p,1,pr),nrow=n,byrow=T)+1

#h<-apply(hap,2,mean)

fid<-paste0("FAM", 1:n)
iid<-paste0("IID", 1:n)
pid<-paste0("PID", 1:n)
mid<-paste0("MID", 1:n)

fid<-1:n
iid<-1:n
pid<-1:n
mid<-1:n

sex<-rep(1,n)
x1<-rnorm(n) #continuous var with effect
x2<-sample(1:5, n, TRUE) #ordered var with effect
x3<-sample(c("aa", "bb_b", "cc#cc", "dddd"), n, TRUE) #categorical var with effect
x4<-rnorm(n) #continuous var without effect
x5<-sample(1:6, n, TRUE) #ordered var without effect
x6<-sample(c("a1a", "b2b", "c3", "dd4"), n, TRUE) #categorical var without effect
x7<-rnorm(n) #continuous var with effect but not observed
x8<-runif(n)
x9<-runif(n)
x10<-sample(1:10, n, TRUE)
x11<-sample(c("a1","a2","a3","a4","a5","a6"), n, TRUE)

x12<-sample(11:20, n, TRUE)
x13<-runif(n,-1,1)
x14<-sample(c("haha", "heyhey", "xixi", "hoho"), n, TRUE)


hap<-data.frame(hap)
colnames(hap)<-paste0("hap", 1:ncol(hap))

p0<-min(p,200)
h1<-hap[,2*(1:p0)-1]-1
h2<-hap[,2*(1:p0)]-1
geno<-h1+h2


#bet<-seq(5,10,length.out=p0)
bet <- runif(p0,-.4,.4)
bet <- bet[order(abs(bet))]
eff<-as.matrix(geno) %*% bet + 1*x1+.5*x2+0.5*(x3=="aa")+1*(x3=="cc#cc")+3*x7
#eff<-as.matrix(geno) %*% bet
eff<-log(1/2/(1-1/2))+eff-mean(eff)
pr<-exp(eff)/(1+exp(eff))
phe<-rbinom(n,1,pr)+1
table(phe)


PED.test<-data.frame(fid,iid,pid,mid,sex,phe,hap)
PED.train<-data.frame(fid,iid,pid,mid,sex,phe,hap[, sel.col])
h1<-hap[,2*(1:p)-1]-1
h2<-hap[,2*(1:p)]-1
geno<-h1+h2
rm(hap)
for(i in 1:100){gc()}

PED.train<-as.matrix(PED.train)
PED.test<-as.matrix(PED.test)

write.table(PED.train, "train.ped",col.names=F,row.names=F,quote=F,append=F)
write.table(PED.test, "test.ped",col.names=F,row.names=F,quote=F,append=F)


chr<-rep(1,p)
snp<-paste0("rs",1:p)
dist<-rep(0,p)
pos<-1:p
MAP.test<-data.frame(chr,snp,dist,pos)
MAP.train<-MAP.test[sel,]
write.table(MAP.train, "train.map", col.names=F,row.names=F,quote=F,append=F)
write.table(MAP.test, "test.map", col.names=F,row.names=F,quote=F,append=F)

#write.table(data.frame(x1,x2,x4,x5,x8,x9), "test.con", col.names=F,row.names=F,quote=F,append=F,sep="\t")
#write.table(data.frame(x3,x6,x10,x11), "test.cat", col.names=F,row.names=F,quote=F,append=F,sep="\t")


con.train<-data.frame(iid,x1,x2,x4,x5,x8,x9)
cat.train<-data.frame(iid,x3,x6,x10,x11)

write.table(con.train, "train.con", col.names=T,row.names=F,quote=F,append=F,sep="\t")
write.table(cat.train, "train.cat", col.names=T,row.names=F,quote=F,append=F,sep="\t")

con.test<-data.frame(iid,x2, x12, x9, x4, x5, x13, x1, x8)
cat.test<-data.frame(iid, x6, x10, x3, x11, x14)
write.table(con.test, "test.con", col.names=T,row.names=F,quote=F,append=F,sep="\t")
write.table(cat.test, "test.cat", col.names=T,row.names=F,quote=F,append=F,sep="\t")



src/main.cpp
///////////////////////////////////////////////////////////////////
//    RANDOM BAMBOO    |     v0.5.5    |      April 29, 2015     //
//---------------------------------------------------------------//
//              (C) 2015 Han Zhang, Yifan Yang                   //
//              GNU General Public License  V3                   //
//---------------------------------------------------------------//
//   For documentation, citation and bug-report instructions:    //
//           http://www.hanzhang.name/softwares/rb               //
//---------------------------------------------------------------//
//                                                               //
//   Update: 0.5.x                                               //
//                                                               //
//     October 9, 2014                                           //
//     (1) remove --neighbor and --searchback introduced         //
//         in v0.4.x. As expected, these two options don't help  //
//         to improve the power due to multiple-comparison issue //
//                                                               //
//     April 16, 2015                                            //
//     (1) add --swt to allow individual spcific sample weights  //
//         in bootstrap                                          //
//                                                               //
//     April 20, 2015                                            //
//     (1) fix a minor bug in function                           //
//         PredictTestingSampleFromMultipleForest(). It leads to //
//         incorrect ntree                                       //
//                                                               //
//     April 24, 2015                                            //
//     (1) add --selcovar to randomly selecte covarites as       //
//         candidate splits for each node. This is the default   //
//         setting in previous version. By adding this option,   //
//         random bamboo now keeps all covariates as candidate   //
//         splits for all nodes. This is reasonable since        //
//         covariates are considered having strong effects on    //
//         the outcome. New option --selcovar alow the users     //
//         to treat covariates equally to SNPs, which is not     //
//         recommended though                                    //
//                                                               //
//     April 29, 2015                                            //
//     (1) add --tidy to mute output files except for bam and    //
//         prd files                                             //
//     (2) fix a minor bug when --pred and --nobam are used      //
//         together                                              //
//                                                               //
///////////////////////////////////////////////////////////////////


#include "config.h"
#include "constant.h"
#include "bamboo.h"


int main(int argc, char **argv){
	
	cout << endl;
	cout << "+------------------------+-------------------+---------------------+" << endl;
	cout << "|     Random Bamboo      |     " << setw(8) << RANDOM_BAMBOO_VERSION << "      |      04/29/2015     |" << endl;
	cout << "+------------------------+-------------------+---------------------+" << endl;
	cout << "|                   (C) 2015 Han Zhang, Yifan Yang                 |" << endl;
	cout << "|                   GNU General Public License, V3                 |" << endl;
	cout << "+------------------------------------------------------------------+" << endl;
	cout << "|     For documentation, citation and bug-report instructions:     |" << endl;
	cout << "|               http://www.hanzhang.name/softwares/rb              |" << endl;
	cout << "+------------------------------------------------------------------+" << endl;
	cout << endl;
	
	char *file = NULL;//f
	char *out = NULL;//o
	char *cont = NULL;//c
	char *cate = NULL;//a
	char *pred = NULL;//p
	char *bam = NULL;//b
	char *trainid = NULL;//y
	char *testid = NULL;//z
	char *snpid = NULL;//S
	char *varid = NULL;//j
	char *samplewt = NULL;//W
	
	int ntree = 1;//t
	int mtry = 0;//m
	int seed = 1;//s
	int max_nleaf = 1000000;//l
	int min_leaf_size = 1;//r
	int imp_measure = 1;//i
	int nthread = sysconf( _SC_NPROCESSORS_ONLN );//h
	//if class_weight < 0, reweight to balance the data
	double class_weight = -1.0;//w
	double cutoff = 1.0;//u
	
	bool flip = false;//g
	bool output_prox = false;//x
	bool output_imp = true;//n
	bool balance = false;//B //if it is turned on and class_weight is not set (< .0), then class_weight by balancing the data (according to case/control ratio in data)
	bool trace = false;//r
	bool output_bamboo = true;//N
	bool keepcovar = true;//k
	bool tidy = false;//T
	
	bool file_spec = false;
	bool out_spec = false;
	bool pred_spec = false;
	bool bam_spec = false;
	bool ntree_spec = false;
	
	int c;
	while(true){
		static struct option long_options[] = {
			{"file", 1, NULL, 'f'},
			{"out", 1, NULL, 'o'},
			{"cont", 1, NULL, 'c'},
			{"cate", 1, NULL, 'a'},
			{"pred", 1, NULL, 'p'},
			{"bam", 1, NULL, 'b'},
			{"trainid", 1, NULL, 'y'},
			{"testid", 1, NULL, 'z'},
			{"snpid", 1, NULL, 'S'},
			{"varid", 1, NULL, 'j'},
			{"ntree", 1, NULL, 't'},
			{"mtry", 1, NULL, 'm'},
			{"seed", 1, NULL, 's'},
			{"maxnleaf", 1, NULL, 'l'},
			{"minleafsize", 1, NULL, 'e'},
			{"imp", 1, NULL, 'i'},
			{"nthread", 1, NULL, 'd'},
			{"classwt", 1, NULL, 'w'},
			{"swt", 1, NULL, 'W'}, 
			{"cutoff", 1, NULL, 'u'},
			{"flip", 0, NULL, 'g'},
			{"prox", 0, NULL, 'x'},
			{"noimp", 0, NULL, 'n'},
			{"balance", 0, NULL, 'B'},
			{"trace", 0, NULL, 'r'},
			{"nobam", 0, NULL, 'N'},
			{"selcovar", 0, NULL, 'R'}, 
			{"tidy", 0, NULL, 'T'}, 
			{"help", 0, NULL, 'h'},
			{"version", 0, NULL, 'v'},
			{0, 0, 0, 0}
		};
		
		//kq
		
		int option_index = 0;
		c = getopt_long(argc, argv, "f:o:c:a:p:b:y:z:S:j:t:m:s:l:e:i:d:w:W:u:gxnBrNRThv", long_options, &option_index);
		
		if(c == -1){
			break;
		}
		
		switch(c){
			case 'f':
				file = optarg;
				file_spec = true;
				break;
			case 'o':
				out = optarg;
				out_spec = true;
				break;
			case 'c':
				cont = optarg;
				break;
			case 'a':
				cate = optarg;
				break;
			case 'p':
				pred = optarg;
				pred_spec = true;
				break;
			case 'b':
				bam = optarg;
				bam_spec = true;
				break;
			case 'y':
				trainid = optarg;
				break;
			case 'z':
				testid = optarg;
				break;
			case 'S':
				snpid = optarg;
				break;
			case 'j':
				varid = optarg;
				break;
			case 't':
				ntree = atoi(optarg);
				ntree_spec = true;
				break;
			case 'm':
				mtry = atoi(optarg);
				if(mtry <= 0){
					cout << "Error: The option --mtry must be a positive integer" << endl;
					return 0;
				}
				break;
			case 's':
				seed = atoi(optarg);
				break;
			case 'l':
				max_nleaf = atoi(optarg);
				if(max_nleaf <= 0){
					cout << "Error: The option --maxnleaf must be a positive integer" << endl;
					return 0;
				}
				break;
			case 'e':
				min_leaf_size = atoi(optarg);
				if(min_leaf_size <= 0){
					cout << "Error: The option --minleafsize must be a positive integer" << endl;
					return 0;
				}
				break;
			case 'i':
				imp_measure = atoi(optarg);
				if(imp_measure != IMP_GINI && imp_measure != IMP_BREIMAN_CUTLER && imp_measure != IMP_LIAW_WIENER){
					cout << "Error: The option --imp only support " << IMP_GINI << " (default), " << IMP_BREIMAN_CUTLER << " and " << IMP_LIAW_WIENER << " right now. Program terminates" << endl;
					return 0;
				}
				if(imp_measure == IMP_NULL){
					imp_measure = 1;
				}
				break;
			case 'd':
				nthread = atoi(optarg);
				if(nthread <= 0){
					cout << "Error: The option --nthread must be a positive integer" << endl;
					return 0;
				}
				
				if(nthread > sysconf( _SC_NPROCESSORS_ONLN )){
					nthread = sysconf( _SC_NPROCESSORS_ONLN );
					cout << "Warning: --nthread is reset to be " << sysconf( _SC_NPROCESSORS_ONLN ) << " (# CPUs)" << endl;
				}
				break;
			case 'w':
				class_weight = atof(optarg);
				if(class_weight <= .0){
					cout << "Error: The option --classwt must be a positive number" << endl;
					return 0;
				}
				break;
			case 'W':
				samplewt = optarg;
				break;
			case 'u':
				cutoff = atof(optarg);
				if(cutoff <= .0){
					cout << "Error: The option --cutoff must be a positive number" << endl;
					return 0;
				}
				break;
			case 'g':
				flip = true;
				break;
			case 'x':
				output_prox = true;
				break;
			case 'n':
				output_imp = false;
				break;
			case 'B':
				balance = true;
				break;
			case 'r':
				trace = true;
				break;
			case 'N':
				output_bamboo = false;
				break;
			case 'R':
				keepcovar = false;
				break;
			case 'T':
				tidy = true;
				break;
			case 'h':
				//print help info here
				return 0;
			case 'v':
				return 0;
			default:
				cout << "Error: Unknow options. Program terminates" << endl;
				exit(1);
		}
	}
	
	
	if(!file_spec){//no plink file is specified
		if(!pred_spec && !bam_spec){
			cout << "Error: The option --file must be specified" << endl;
			return 0;
		}else if(pred_spec && !bam_spec){
			cout << "Error: The option --bam must be specified for prediction purpose" << endl;
			return 0;
		}else if(!pred_spec && bam_spec){//print the information of the specified model
			BAMBOO bb(bam, nthread);
			bb.PrintModelInfo();
			return 0;
		}else{
			;//good
		}
	}else{//train model from specified data
		bam = NULL;
		bam_spec = false;
	}
	
	if(!out_spec){//if the name of output file is not specified
		if(file_spec && pred_spec){//fit model and predict new data
			cout << "Error: When options --file and --pred are specified simultaneouslythe option --out must be spcified by the user" << endl;
			exit(1);
		}else if(file_spec && !pred_spec){//fit model but don't predict new data
			out = file;
		}else if(!file_spec && pred_spec){//predict new data by given model
			out = pred;
		}else{//impossible
			cout << "Error: debug" << endl;
			exit(1);
		}
	}
	
	if(balance && class_weight > .0){
		cout << "Error: Please don't use the option --balance with option --classwt simultaneously" << endl;
		return 0;
	}
	
	//if balance = true, the program will not check the value of class_weight and will determine it automatically
	//so don't worry about the following. If balance = false and class_weight is not set by the user (class_weight < .0), 
	//then set it as 1.0 (means no weight). So it is important to initialize class_weight as -1.0 at the very beginning
	if(class_weight <= .0){
		class_weight = 1.0;
	}
	
	
	if(!file_spec && pred_spec){//predicting testing data (--pred) using models specified in --bam
		if(!ntree_spec){
			ntree = -1;
		}
		BAMBOO bb (out, pred, bam, testid, nthread, ntree);
		bb.PredictTestingSample();
		return 0;
	}else{//training model, and then predicting testing data if --pred is on
		BAMBOO bb (file, out, cont, cate, pred, trainid, testid, snpid, samplewt, ntree, mtry, max_nleaf, 
		min_leaf_size, imp_measure, seed, nthread, class_weight, cutoff, flip, output_prox, output_imp, 
		output_bamboo, keepcovar, tidy, balance, trace);
		bb.GrowForest();
		bb.PredictTestingSample();
		return 0;
	}
	
	return 0;
	
}
